File: server/BudgetBoard-Database/Models/Goal.cs
﻿namespace BudgetBoard.Database.Models;

public class Goal
{
    public Guid ID { get; set; }
    public string Name { get; set; } = string.Empty;
    public DateTime? CompleteDate { get; set; }
    public decimal Amount { get; set; } = 0.0M;
    public decimal? InitialAmount { get; set; }
    public decimal? MonthlyContribution { get; set; }
    public ICollection<Account> Accounts { get; set; } = new List<Account>();
    public required Guid UserID { get; set; }
    public ApplicationUser? User { get; set; } = null!;
}

File: client/src/models/goal.ts
import { IAccountResponse } from "./account";

export interface IGoalCreateRequest {
  name: string;
  completeDate: string | null;
  amount: number;
  applyExistingBalanceTowardsGoal: boolean;
  monthlyContribution: number | null;
  accountIds: string[];
}

export interface IGoalUpdateRequest {
  id: string;
  name?: string;
  completeDate?: string | null;
  amount?: number;
  monthlyContribution?: number | null;
}

export interface IGoalResponse {
  id: string;
  name: string;
  completeDate: Date;
  isCompleteDateEditable: boolean;
  amount: number;
  initialAmount: number;
  monthlyContribution: number;
  isMonthlyContributionEditable: boolean;
  monthlyContributionProgress: number;
  interestRate: number | null;
  completed: Date | null;
  percentComplete: number;
  accounts: IAccountResponse[];
  userID: string;
}

export enum GoalType {
  None = "",
  SaveGoal = "saveGoal",
  PayGoal = "payGoal",
}

export enum GoalCondition {
  TimedGoal = "timedGoal",
  MonthlyGoal = "monthlyGoal",
}

export enum GoalTarget {
  TargetBalanceGoal = "targetBalanceGoal",
  TargetAmountGoal = "targetAmountGoal",
}
